### ISSUE TITLE
Gracetime not enforced in all cases

### ISSUE THREAD

**Issue by stefanschneider** [cloudfoundry-incubator/garden#40]
We saw an issue with gracetime in graden-windows. The grace time is not enforced on containers that had running processes. I seams to be the same for garden-linux.

To reproduce on bosh-lite + cf + diego:
```
# start a diego app
bosh ssh cell_z1 0
monit stop rep
# after 5 minutes (the default gracetime) the app not not get stopped
```

This is a suggestion for a fix that fixes the issue in my env:
https://github.com/stefanschneider/garden/commit/edd2551ae01174e2f11718f450b9ff86f4c6fed2

**Comment by julz** [comment]
Hi @stefanschneider I _think_ what you're describing is the intended behaviour, but I could have misunderstood so let me try to make sure we're both talking about the same thing :). GraceTime is the amount of time a container is allowed to survive if no client interacts with it. When diego runs an app it keeps open a stream to get that process's output and exit status, and that (intentionally) keeps the GraceTime counter alive (because we don't want diego apps to be killed if diego is still waiting for them to exit normally). GraceTime is used to garbage collect apps that no client has asked about (including being connected to a `garden.Process` waiting for an exit status/streaming io) for the given time. 

Are you seeing something consistent with the above (i.e. the app is kept alive while diego knows about it, but is cleaned up after the grace time if diego is killed and doesn't come back after that amount of time) or are you seeing something that's _not_ consistent with that (i.e. the app is not killed even if all clients have gone away and stopped waiting for an exit status)?

**Comment by stefanschneider** [comment]
The issues is that when the rep stops (the only client that talks to the garden server) the gracetime is not enforced. In my understand of the gracetime is that when there is no client all containers should be reaped after the gracetime period.

**Comment by stefanschneider** [comment]
In case of dea_ng + warden, when the dea_ng and other clients (dea_logging_agnet) stop, the gracetime will be enforced by the warden server.

**Comment by julz** [comment]
yeah, you're absolutely right, if no client is talking the container should be killed after the grace time. Are you sure the executor (as opposed to the rep) is not still keeping a connection open (I may not be up to date on what talks to what!)?

**Comment by stefanschneider** [comment]
The executor lives in the rep process space. I am sure about that.

**Comment by julz** [comment]
ok, then that definitely sounds like a bug. Since it looks like you already have a fix above would you be interested in adding a test case and making a pull request? otherwise I'll create a bug in our tracker..

**Comment by stefanschneider** [comment]
yeah, of course. i'll be pairing with @mhoran to also add some test for this. 
Can you tell us if the approach from this patch is the right way to move forward:
https://github.com/stefanschneider/garden/commit/edd2551ae01174e2f11718f450b9ff86f4c6fed2

**Comment by julz** [comment]
yeah reading it through I think I see what's going on now and the fix makes a lot of sense. You may want to use a `chan struct{}` rather than a `chan bool` and then `close` it rather than sending a bool on it (since you're really using it as a broadcast mechanism `chan struct{}` with a `close` is probably slightly more idiomatic, and avoids any possibility of getting deadlocked since it won't block). (also thanks! :))

**Comment by garden-gnome** [comment]
Hi there!

We use Pivotal Tracker to provide visibility into what our team is working on. A story for this issue has been automatically created.

The current status is as follows:

* [ ] [#109274424](https://www.pivotaltracker.com/story/show/109274424) Gracetime not enforced in all cases


This comment, as well as the labels on the issue, will be automatically updated as the status in Tracker changes.

**Comment by goonzoid** [comment]
@stefanschneider @julz was this fixed by PR #43?
